core/src/pageId.rs

```diff
@@ -0,0 +1,131 @@
+//! TODO
+
+use crate::{page::DEPTH, trie::KeyPath};
+use bitvec::prelude::*;
+
+pub type PageId = [u8; 32];
+
+pub const ROOT_PAGE_ID: [u8; 32] = [0; 32];
+
+/// Each Page is identified by a unique PageId,
+/// Root Page has id 0, its leftmost child 1 and so on.
+///
+/// There are 2^250 possible pages, so 250 bits are needed to represent all of them.
+///
+/// An array of bytes will be used as a container for the PageId,
+/// bytes will be treated as a 256-bit big-endian integer.
+///
+/// (TODO: treating them as little-endian would simply a lot offsets)
+///
+/// Given a KeyPath, it will be divided into chunks of 6 bits (sextet), forming
+/// an array of sextets. Each PageId will be constructed in the following manner:
+///
+/// page_ids[i] = (prev_page_id << d) + int(sextets[i]) + 1
+///
+/// The following struct implements an Iterator of PageIds over a KeyPath
+/// to lazily construct them as needed
+struct PageIdsIterator {
+    key_path: KeyPath,
+    level: u8, // there will be at most 42 levels
+    next_page_id: PageId,
+}
+
+// TODO: This probably is not the best way to construct the iterator
+pub fn get_page_ids(key_path: KeyPath) -> impl Iterator<Item = PageId> {
+    PageIdsIterator {
+        key_path,
+        level: 0,
+        next_page_id: ROOT_PAGE_ID,
+    }
+}
+
+impl Iterator for PageIdsIterator {
+    type Item = PageId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let prev_page_id = self.next_page_id;
+
+        // The last bits will not be used as a sextet to identify a page.
+        // If sextets are finished, then there are no more pages to load.
+        let sextet = self
+            .key_path
+            .view_bits::<Msb0>()
+            .chunks_exact(DEPTH)
+            .nth(self.level as usize)?;
+
+        //#[cfg(feature = "std")]
+        //println!("level: {}, sextet: {:?}", self.level, sextet);
+
+        let next_page_id: &mut BitSlice<u8, Msb0> = self.next_page_id.view_bits_mut();
```

Comment:
Using Msb0 for the KeyPath is perfect, to maintain consistency I tried to use the same convention for the PageId as well. It's feasible but the indices used are not very convenient in this case, especially when working mainly with the least significant byte. Using an Lsb0 convention could simplify things.

The handling of PageId is mainly restricted to this Iterator, so there are no other places where one must remember that the PageId is a little-endian integer. What do you think? @rphmeier @pepyakin should we keep Msb0 also here?

Reply:
How does Lsb0 make it easier?